File: backend/app/schemas/suivi_sol_vivant.py
"""Schemas Pydantic pour le Suivi des Sols Vivants"""
from pydantic import BaseModel, ConfigDict
from typing import Optional, List
from datetime import date


# ── Réamendements ──────────────────────────────────────────────────────────────

class SuiviReamendementBase(BaseModel):
    id_recette_reamend: Optional[int]  = None
    date_application:   Optional[date] = None
    notes:              Optional[str]  = None


class SuiviReamendementCreate(SuiviReamendementBase):
    pass


class SuiviReamendementRead(SuiviReamendementBase):
    id_suivi_reamend:   int
    nom_recette_reamend: Optional[str]  = None
    cout_estime:         Optional[float]= None
    model_config = ConfigDict(from_attributes=True)


# ── Arrosages ─────────────────────────────────────────────────────────────────

class SuiviArrosageBase(BaseModel):
    id_recette_engrais:  Optional[int]  = None
    volume_eau_l:        Optional[float]= None
    date_application:    Optional[date] = None
    notes:               Optional[str]  = None


class SuiviArrosageCreate(SuiviArrosageBase):
    pass


class SuiviArrosageRead(SuiviArrosageBase):
    id_suivi_arrosage:   int
    nom_recette_arrosage: Optional[str]  = None
    cout_estime:          Optional[float]= None
    model_config = ConfigDict(from_attributes=True)


# ── TCO ───────────────────────────────────────────────────────────────────────

class SuiviTCOBase(BaseModel):
    id_recette_tco:  Optional[int]  = None
    volume_applique: Optional[float]= None
    date_application: Optional[date]= None
    notes:           Optional[str]  = None


class SuiviTCOCreate(SuiviTCOBase):
    pass


class SuiviTCORead(SuiviTCOBase):
    id_suivi_tco:   int
    nom_recette_tco: Optional[str] = None
    model_config = ConfigDict(from_attributes=True)


# ── Fermentation ──────────────────────────────────────────────────────────────

class SuiviFermentationBase(BaseModel):
    id_recette_ferm: Optional[int]  = None
    volume_applique: Optional[float]= None
    date_application: Optional[date]= None
    notes:           Optional[str]  = None


class SuiviFermentationCreate(SuiviFermentationBase):
    pass


class SuiviFermentationRead(SuiviFermentationBase):
    id_suivi_ferm:    int
    nom_recette_ferm: Optional[str] = None
    model_config = ConfigDict(from_attributes=True)


# ── Cultures ──────────────────────────────────────────────────────────────────

class SuiviCultureBase(BaseModel):
    description: Optional[str]  = None
    date_debut:  Optional[date] = None
    date_fin:    Optional[date] = None
    notes:       Optional[str]  = None


class SuiviCultureCreate(SuiviCultureBase):
    pass


class SuiviCultureRead(SuiviCultureBase):
    id_suivi_culture: int
    model_config = ConfigDict(from_attributes=True)


# ── Suivi principal ───────────────────────────────────────────────────────────

class SuiviSolVivantBase(BaseModel):
    nom_pot:         str
    id_materiel:     Optional[int]  = None
    id_recette_lso:  Optional[int]  = None
    volume_pot_l:    Optional[float]= None
    date_preparation: Optional[date]= None
    commentaires:    Optional[str]  = None


class SuiviSolVivantCreate(SuiviSolVivantBase):
    reamendements: List[SuiviReamendementCreate] = []
    arrosages:     List[SuiviArrosageCreate]     = []
    tcos:          List[SuiviTCOCreate]          = []
    fermentations: List[SuiviFermentationCreate] = []
    cultures:      List[SuiviCultureCreate]      = []


class SuiviSolVivantUpdate(BaseModel):
    nom_pot:         Optional[str]  = None
    id_materiel:     Optional[int]  = None
    id_recette_lso:  Optional[int]  = None
    volume_pot_l:    Optional[float]= None
    date_preparation: Optional[date]= None
    commentaires:    Optional[str]  = None
    reamendements:   Optional[List[SuiviReamendementCreate]] = None
    arrosages:       Optional[List[SuiviArrosageCreate]]     = None
    tcos:            Optional[List[SuiviTCOCreate]]          = None
    fermentations:   Optional[List[SuiviFermentationCreate]] = None
    cultures:        Optional[List[SuiviCultureCreate]]      = None


class SuiviSolVivantRead(SuiviSolVivantBase):
    id_suivi:          int
    nom_recette_lso:   Optional[str]  = None
    nom_materiel:      Optional[str]  = None
    cout_lso_estime:   Optional[float]= None
    cout_total_estime: Optional[float]= None
    reamendements:     List[SuiviReamendementRead]  = []
    arrosages:         List[SuiviArrosageRead]      = []
    tcos:              List[SuiviTCORead]           = []
    fermentations:     List[SuiviFermentationRead]  = []
    cultures:          List[SuiviCultureRead]       = []
    model_config = ConfigDict(from_attributes=True)

File: frontend/src/api/suiviSolVivant.ts
import client from './client'

export interface SuiviReamendement {
  id_suivi_reamend:     number
  id_recette_reamend?:  number
  date_application?:    string
  notes?:               string
  nom_recette_reamend?: string
  cout_estime?:         number
}

export interface SuiviArrosage {
  id_suivi_arrosage:     number
  id_recette_engrais?:   number
  volume_eau_l?:         number
  date_application?:     string
  notes?:                string
  nom_recette_arrosage?: string
  cout_estime?:          number
}

export interface SuiviTCO {
  id_suivi_tco:    number
  id_recette_tco?: number
  volume_applique?: number
  date_application?: string
  notes?:          string
  nom_recette_tco?: string
}

export interface SuiviFermentation {
  id_suivi_ferm:    number
  id_recette_ferm?: number
  volume_applique?: number
  date_application?: string
  notes?:           string
  nom_recette_ferm?: string
}

export interface SuiviCulture {
  id_suivi_culture: number
  description?:     string
  date_debut?:      string
  date_fin?:        string
  notes?:           string
}

export interface SuiviSolVivant {
  id_suivi:           number
  nom_pot:            string
  id_materiel?:       number
  id_recette_lso?:    number
  volume_pot_l?:      number
  date_preparation?:  string
  commentaires?:      string
  nom_recette_lso?:   string
  nom_materiel?:      string
  cout_lso_estime?:   number
  cout_total_estime?: number
  reamendements:      SuiviReamendement[]
  arrosages:          SuiviArrosage[]
  tcos:               SuiviTCO[]
  fermentations:      SuiviFermentation[]
  cultures:           SuiviCulture[]
}

export type SuiviSolVivantCreate = Omit<SuiviSolVivant, 'id_suivi' | 'nom_recette_lso' | 'nom_materiel' | 'cout_lso_estime' | 'cout_total_estime'> & {
  reamendements: Omit<SuiviReamendement, 'id_suivi_reamend' | 'nom_recette_reamend' | 'cout_estime'>[]
  arrosages:     Omit<SuiviArrosage, 'id_suivi_arrosage' | 'nom_recette_arrosage' | 'cout_estime'>[]
  tcos:          Omit<SuiviTCO, 'id_suivi_tco' | 'nom_recette_tco'>[]
  fermentations: Omit<SuiviFermentation, 'id_suivi_ferm' | 'nom_recette_ferm'>[]
  cultures:      Omit<SuiviCulture, 'id_suivi_culture'>[]
}

export const suiviSolVivantAPI = {
  getAll:  ()                                                => client.get<SuiviSolVivant[]>('/suivi-sols-vivants/'),
  getById: (id: number)                                      => client.get<SuiviSolVivant>(`/suivi-sols-vivants/${id}`),
  create:  (data: SuiviSolVivantCreate)                      => client.post<SuiviSolVivant>('/suivi-sols-vivants/', data),
  update:  (id: number, data: Partial<SuiviSolVivantCreate>) => client.put<SuiviSolVivant>(`/suivi-sols-vivants/${id}`, data),
  delete:  (id: number)                                      => client.delete(`/suivi-sols-vivants/${id}`),
  addReamendement: (id: number, data: object)  => client.post<SuiviSolVivant>(`/suivi-sols-vivants/${id}/reamendements`, data),
  addArrosage:     (id: number, data: object)  => client.post<SuiviSolVivant>(`/suivi-sols-vivants/${id}/arrosages`, data),
  addTCO:          (id: number, data: object)  => client.post<SuiviSolVivant>(`/suivi-sols-vivants/${id}/tcos`, data),
  addFermentation: (id: number, data: object)  => client.post<SuiviSolVivant>(`/suivi-sols-vivants/${id}/fermentations`, data),
  addCulture:      (id: number, data: object)  => client.post<SuiviSolVivant>(`/suivi-sols-vivants/${id}/cultures`, data),
}
